tests/simple_regular_exchange.cpp

```diff
@@ -0,0 +1,243 @@
+    res = res && check(field, dims);
+    return res;
+}
+
+template<typename Context, typename Pattern, typename Domains>
+bool run(Context& context, const Pattern& pattern, const Domains& domains, const arr& dims)
+{
+    bool res = true;
+    // fields
+    auto raw_field_a = allocate_field();
+    auto raw_field_b = allocate_field();
+    auto field_a     = fill(wrap_field(raw_field_a, domains[0]));
+    auto field_b     = fill(wrap_field(raw_field_b, domains[1]));
+    // get a communcator
+    auto comm = context.get_communicator(context.get_token());
+    
+    // general exchange
+    // ================
+    auto co = make_communication_object<Pattern>(comm);
+    auto h0 = co.exchange(pattern(field_a), pattern(field_b));
+    // ...
+    h0.wait();
+
+    // check fields
+    res = res && check(field_a, dims);
+    res = res && check(field_b, dims);
+
+    // reset fields
+    reset(field_a);
+    reset(field_b);
+
+    // bulk exchange (rma)
+    // ===================
+    auto bco = bulk_communication_object<structured::rma_range_generator, Pattern, decltype(field_a)>(comm);
+    bco.add_field(pattern(field_a));
+    bco.add_field(pattern(field_b));
+    auto h1 = bco.exchange();
+    // ...
+    h1.wait();
+
+    // check fields
+    res = res && check(field_a, dims);
+    res = res && check(field_b, dims);
+    return res;
+}
+
+void sim(bool multi_threaded)
+{
+    // make a context from mpi world and number of threads
+    auto context_ptr = factory::create(multi_threaded? 2 : 1, MPI_COMM_WORLD);
+    auto& context    = *context_ptr;
+    // 2D domain decomposition
+    arr dims{0,0}, coords{0,0};
+    MPI_Dims_create(context.size(), 2, dims.data());
+    coords[1] = context.rank()/dims[0];
+    coords[0] = context.rank() - coords[1]*dims[0];
+    // make 2 domains per rank
+    std::vector<domain> domains{
+        make_domain(context.rank(), 0, coords),
+        make_domain(context.rank(), 1, coords)};
+    // make halo generator
+    halo_gen gen{arr{0,0}, arr{dims[0]*DIM-1,dims[1]*DIM-1}, halos, periodic};
+    // create a pattern for communication
+    auto pattern = make_pattern<structured::grid>(context, gen, domains);
+    // run
+    bool res = true;
+    if (multi_threaded)
+    {
+        auto run_fct = [&context,&pattern,&domains,&dims](int id)
+            { return run(context, pattern, domains, dims, id); };
+        auto f1 = std::async(std::launch::async, run_fct, 0);
+        auto f2 = std::async(std::launch::async, run_fct, 1);
+        res = res && f1.get();
+        res = res && f2.get();
+    }
+    else
+    {
+        res = res && run(context, pattern, domains, dims);
+    }
+    // reduce res
+    bool all_res = false;
+    MPI_Reduce(&res, &all_res, 1, MPI_C_BOOL, MPI_LAND, 0, MPI_COMM_WORLD);
+    if (context.rank() == 0)
+    {
+        EXPECT_TRUE(all_res);
+    }
+}
+
+TEST(simple_regular_exchange, single)
+{
+    sim(false);
+}
+
+TEST(simple_regular_exchange, multi)
+{
+    sim(false);
```

Comment:
```suggestion
    sim(true);
```
?

Reply:
obviously! I will also add the gpu test now